sql-plugin/src/main/scala/org/apache/spark/sql/rapids/HashFunctions.scala

```diff
@@ -108,6 +108,59 @@ case class GpuXxHash64(children: Seq[Expression], seed: Long) extends GpuHashExp
   }
 }
 
+object XxHash64Utils {
+  /**
+   * Convert map to list of struct.
+   * Note: Do not support UserDefinedType and other unregular types.
+   * Do not retain the nullable and other info
+   */
+  private def flatMap(inputType: DataType): DataType = {
+    inputType match {
+      case mapType: MapType =>
+        ArrayType(StructType(Array(
+          StructField("key", flatMap(mapType.keyType)),
+          StructField("value", flatMap(mapType.valueType))
+        )))
+      case arrayType: ArrayType => ArrayType(flatMap(arrayType.elementType))
+      case structType: StructType =>
+        StructType(structType.map(f => StructField(f.name, flatMap(f.dataType))).toArray)
+      case nullType: NullType => nullType
+      case atomicType: AtomicType => atomicType
+      case other => throw new RuntimeException(s"Unsupported type: $other")
+    }
+  }
+
+  /**
+   * Compute the max stack size that the flattenType will use.
+   * @param flattenType should be fatten type generated by function `flatMap`
+   * @return max stack size
+   */
+  private def computeMaxStackSizeForFlatten(flattenType: DataType): Int = {
+    flattenType match {
+      case ArrayType(c: StructType, _) => 1 + computeMaxStackSizeForFlatten(c)
+      case ArrayType(c: DataType, _) => computeMaxStackSizeForFlatten(c)
+      case st: StructType =>
+        1 + st.map(f => computeMaxStackSizeForFlatten(f.dataType)).max
+      case _ => 1 // primitive types
```

Comment:
Perhaps we don't need the `flatMap` function and can directly compute the depth of the map type using 
```
case mt: MapType => 
    2 + math.max(computeMaxStackSizeForFlatten(mt.keyType), computeMaxStackSizeForFlatten(mt.valueType))
```

Reply:
+1 for not rewriting the data type to check the depth.

Reply:
Thanks, let's post a follow-up PR to do the improvement

Reply:
@res-life why? It takes 5 mins and I want the fallback message updated too?

Reply:
Updated.